code/go/internal/validator/folder_item_content.go

```diff
@@ -9,59 +9,115 @@ import (
-	itemData, err := fs.ReadFile(fsys, itemPath)
+func loadItemSchema(fsys fs.FS, path string, contentType *spectypes.ContentType) ([]byte, error) {
+	data, err := fs.ReadFile(fsys, path)
 	if err != nil {
-		return nil, errors.Wrap(err, "reading item file failed")
+		return nil, ve.ValidationErrors{errors.Wrap(err, "reading item file failed")}
+	}
+	if contentType != nil && contentType.MediaType == "application/x-yaml" {
+		return convertYAMLToJSON(data)
 	}
+	return data, nil
+}
 
-	// There might be a situation in which we'd like to keep a directory without any file content under version control.
-	// Usually it can be solved by adding the .empty file. There is no media type defined for this file.
-	// It's expected of the file with media type defined not to be empty - the file can be marked as non-required.
-	if len(itemData) == 0 && mediaType != "" {
-		return nil, errors.New("file is empty, but media type is defined")
+func validateContentType(fsys fs.FS, path string, contentType spectypes.ContentType) error {
+	switch contentType.MediaType {
+	case "application/x-yaml":
+		v, _ := contentType.Params["require-document-dashes"]
+		requireDashes := (v == "true")
+		if requireDashes {
+			err := validateYAMLDashes(fsys, path)
+			if err != nil {
+				return err
+			}
+		}
+	case "application/json":
+	case "text/markdown":
+	case "text/plain":
+	default:
+		return fmt.Errorf("unsupported media type (%s)", contentType)
 	}
+	return nil
+}
 
-	if mediaType == "" {
-		return itemData, nil // no item's schema defined
+func validateYAMLDashes(fsys fs.FS, path string) error {
+	f, err := fsys.Open(path)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+
+	dashes := []byte("---\n")
+	b := make([]byte, len(dashes))
+	_, err = f.Read(b)
+	if err != nil {
+		return err
 	}
 
-	basicMediaType, params, err := mime.ParseMediaType(mediaType)
+	if !bytes.Equal(dashes, b) {
+		return errors.New("document dashes are required (start the document with '---')")
+	}
+	return nil
+}
+
+func validateContentTypeSize(fsys fs.FS, path string, contentType spectypes.ContentType, limits commonSpecLimits) error {
+	info, err := fs.Stat(fsys, path)
 	if err != nil {
-		return nil, errors.Wrapf(err, "invalid media type (%s)", mediaType)
+		return err
+	}
+	size := spectypes.FileSize(info.Size())
+	if size <= 0 {
+		return errors.New("file is empty, but media type is defined")
 	}
 
-	switch basicMediaType {
+	var sizeLimit spectypes.FileSize
+	switch contentType.MediaType {
 	case "application/x-yaml":
-		// TODO Determine if special handling of `---` is required (issue: https://github.com/elastic/package-spec/pull/54)
-		if v, _ := params["require-document-dashes"]; v == "true" && !bytes.HasPrefix(itemData, []byte("---\n")) {
-			return nil, errors.New("document dashes are required (start the document with '---')")
-		}
+		sizeLimit = limits.ConfigurationSizeLimit
+	}
+	if sizeLimit > 0 && size > sizeLimit {
```

Comment:
Question: empty YAML file. Is it a valid file :) ?

Reply:
There is a check some lines above that makes validation to fail on empty files that have a content type defined, I brought this from previous implementation, though I think that there are no tests for this.

The truth is that if a file is empty, it cannot satisfy the condition of having a content type :)

Reply:
It seems to be more of a life-death discussion if the empty file is valid JSON/YAML :) Never mind, wanted to double-check if there isn't any bug around.